library/std/src/sys/pal/common/small_c_string.rs

```diff
@@ -17,20 +17,32 @@ const NUL_ERR: io::Error =
 #[inline]
 pub fn run_path_with_cstr<T, F>(path: &Path, f: F) -> io::Result<T>
 where
-    F: FnOnce(&CStr) -> io::Result<T>,
+    F: FnMut(&CStr) -> io::Result<T>,
 {
     run_with_cstr(path.as_os_str().as_encoded_bytes(), f)
 }
 
 #[inline]
-pub fn run_with_cstr<T, F>(bytes: &[u8], f: F) -> io::Result<T>
+pub fn run_with_cstr<T, F>(bytes: &[u8], mut f: F) -> io::Result<T>
```

Comment:
Why are you not using `dyn` here as well (and in `run_path`)?

Reply:
Changes the call signature for users, as they would have to pass `&mut`, and I'm trying to keep PRs as small as possible.

Reply:
Given that this extraction does make the code a bit less pretty (imo) and that there are only like 10 callers, I think it makes sense to change `run_with_cstr` too. I'm happy to review/approve this PR then.

Reply:
Done